### DISCUSSION TITLE
Excessive Restarts in Pocketbase with Kubernetes and Persistent Volume

### DISCUSSION THREAD

**Discussion by cavecafe** [pocketbase/pocketbase#4981] [Q&A] [answered]
### Excessive Restarts in Pocketbase with Kubernetes and Persistent Volume

I am experiencing excessive restarts in my Pocketbase deployment running on Kubernetes with a Persistent Volume (PV). The logs indicate frequent restarts triggered by the 'debounceTimer', which is triggered by changes to the /pb_hooks/logs.db-shm file.

### Specifics:
* Pocketbase version: v0.22.12
* Kubernetes version: v1.28.6
* Persistent Volume type: (M.2 SSD, only assigned 1GB, YAML attached for PV and PVC)
* Dockerfile: (in below)

### Observations:
* The logs show repeated entries like:
```log
File /pb_hooks/logs.db-shm changed, restarting...
2024/05/24 00:55:40 Server started at http://0.0.0.0:8090

... (repeated every a few milli-seconds)

```
![pocketbase-logs](https://github.com/pocketbase/pocketbase/assets/758641/76fd11d0-d91d-4a0e-a137-6d653aa8eaf4)

* I am not a Go developer, but I have provided the relevant code snippet for reference in below. The code snippet in the 'debounceTimer' function indicates that the application restarts upon detecting changes to the /pb_hooks/logs.db-shm file.(No idea who and why this file get changed, and why this require to restart the entire server)

![pocketbase-code-for-log](https://github.com/pocketbase/pocketbase/assets/758641/57d30cb9-d7e9-4cc2-89e9-e76c8ab9a351)

### Questions:

* Is the frequent restarting of the application expected behavior due to changes in the /pb_hooks/logs.db-shm file?
* If not, how can I prevent these excessive restarts?
* Are there any configuration options or settings that I can adjust to address this issue?

I would greatly appreciate any assistance from the community in understanding and resolving this issue. Please let me know if you require any further information or clarification.

Thank you for your time and support.

### Additional Information:
* I have also included the Dockerfile and the YAML configuration for the Persistent Volume and Persistent Volume Claim. PV, PVC are created already before Docker run in K8s
```dockerfile
FROM alpine:latest as downloader

RUN wget --tries=10 https://github.com/pocketbase/pocketbase/releases/download/v0.22.12/pocketbase_0.22.12_linux_amd64.zip \
    && unzip pocketbase_0.22.12_linux_amd64.zip \
    && chmod +x /pocketbase

FROM alpine:latest
RUN apk update && apk add ca-certificates && rm -rf /var/cache/apk/*

EXPOSE 8090

COPY --from=downloader /pocketbase /usr/local/bin/pocketbase

RUN addgroup --system --gid 1001 pbadmin
RUN adduser --system --uid 1001 --ingroup pbadmin pbadmin

WORKDIR /pocketbase

USER pbadmin

ENTRYPOINT ["/usr/local/bin/pocketbase", "serve", "--http=0.0.0.0:8090", \
    "--dir=/pocketbase/pb_data", \
    "--publicDir=/pocketbase/pb_public", \
    "--hooksDir=/pocketbase/pb_hooks"]
```
* PV, PVC is created ahead of docker image deployed, and connect well with Docker run. 
  
```yaml
apiVersion: v1
kind: PersistentVolume
metadata:
  annotations:
    longhorn.io/volume-scheduling-error: ""
    pv.kubernetes.io/bound-by-controller: "yes"
  creationTimestamp: "2024-05-24T00:59:17Z"
  finalizers:
  - kubernetes.io/pv-protection
  - external-attacher/driver-longhorn-io
  name: pocketbase
  resourceVersion: "18516187"
  uid: a28b0109-8f94-4a3e-abb8-9c0ca6949ae8
spec:
  accessModes:
  - ReadWriteOnce
  capacity:
    storage: 1Gi
  claimRef:
    apiVersion: v1
    kind: PersistentVolumeClaim
    name: pocketbase
    namespace: pocketbase
    resourceVersion: "18515555"
    uid: 0c4b2c3e-9d4e-4392-b84f-648abd9295e7
  csi:
    driver: driver.longhorn.io
    fsType: ext4
    volumeAttributes:
      diskSelector: ""
      nodeSelector: ""
      numberOfReplicas: "1"
      staleReplicaTimeout: "20"
    volumeHandle: pocketbase
  persistentVolumeReclaimPolicy: Retain
  storageClassName: pocketbase
  volumeMode: Filesystem
status:
  phase: Bound
---
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  annotations:
    pv.kubernetes.io/bind-completed: "yes"
  creationTimestamp: "2024-05-24T00:59:17Z"
  finalizers:
  - kubernetes.io/pvc-protection
  name: pocketbase
  namespace: pocketbase
  resourceVersion: "18515559"
  uid: 0c4b2c3e-9d4e-4392-b84f-648abd9295e7
spec:
  accessModes:
  - ReadWriteOnce
  resources:
    requests:
      storage: 1Gi
  storageClassName: pocketbase
  volumeMode: Filesystem
  volumeName: pocketbase
status:
  accessModes:
  - ReadWriteOnce
  capacity:
    storage: 1Gi
  phase: Bound
```

**Comment by thangngoc89** [comment]
Not sure why logs.db-shm ended up in pb_hooks but it’s supposed to be in pb-data.

**Reply by cavecafe** [reply]
Thanks for response @thangngoc89, 
If I remember correctly, even though I don't have any additional setup, all three directories pb_data, pb_public, pb_hooks has their own logs.db-shm files. So, I'm kind of being confused now what you mean by that.

**Reply by thangngoc89** [reply]
@cavecafe pb_public and pb_hooks shouldn’t have any db-shm files. The code you mentioned restarts pocketbase when there are changes inside pb_hooks

**Reply by ganigeorgiev** [reply, marked as answer]
> ... all three directories pb_data, pb_public, pb_hooks has their own logs.db-shm files. 

@cavecafe No. As mentioned above there shouldn't be db files in the `pb_hooks` directory, only in `pb_data`.

The app restarts automatically **only** when there are changes in `pb_hooks`. If you want to disable the watcher you could start the process with `--hooksWatch=false`. But this will not fix your actual issue.

How did you end up in the situation I don't know. Either you are not sharing everything (_eg. you have a improperly configured replication setup with Litestream, Marmot, etc._) or there is something wrong with the PVC setup and its mounting point (you usually need a volume only for the `pb_data`; `pb_hooks` and `pb_public` can be safely copied directly into the image; For example Dockerfile you can check [Using Docker](https://pocketbase.io/docs/going-to-production/#using-docker)).

**Comment by cavecafe** [comment]
Thanks for the quick responses @ganigeorgiev @thangngoc89, 

### Resolution: 

Following the suggestion by @ganigeorgiev, I applied the --hooksWatch=false flag when starting Pocketbase. This successfully prevented the excessive restarts. Thanks!


### Additional Notes:

* I 'guess' the restarts were occurring because Pocketbase was detecting changes to the 'pre-existing' files in the /pb_hooks directory, which was mounted as a Persistent Volume ahead of Docker image run. In my Dockerfile for K8s:
```bash 
./pocketbase serve http://localhost:8090 \ 
   --dir=/pocketbase/pb_data \ 
   --publicDir=/pocketbase/pb_public \ 
   --hooksDir=/pocketbase/pb_hooks
```
And, the folders are like these:
![pocketbase-k8s-storages-ls](https://github.com/pocketbase/pocketbase/assets/758641/3e963700-7ce4-412c-8ec8-5d6b5f69a763)

* I have observed a difference in behavior between my local macOS environment and the Kubernetes deployment. On macOS, Pocketbase does not create the pb_public and pb_hooks directories automatically even though I use the same arguments. 

```bash
./pocketbase serve http://localhost:8090 \
   --dir=pb_data \ 
   --publicDir=pb_public \
   --hooksDir=pb_hooks
```
![pocketbase_0 22 12_darwin_arm64_2024-05-24](https://github.com/pocketbase/pocketbase/assets/758641/060f1649-5ab8-443b-97b2-72ff3dd9f410)

### Conclusion for my uses
* Simply let 'pocketbase' create these folders, just prepare base directory /pocketbase/ only in the PV. 
* I did not try it yet. I'll post the result soon. :)

It is quick and helpful, I really appreciate it! 
Happy Coding folks!
Thomas

**Reply by ganigeorgiev** [reply]
>  I applied the --hooksWatch=false flag when starting Pocketbase. This successfully prevented the excessive restarts.

Again, as mentioned in my comment, while this may "solve" the restart issue you are experiencing, it is not really a solution for your problem because there shouldn't be generated db files in the `pb_hooks` directory in the first place. Something else is going on in your environment and I'd recommend to investigate it as it could cause other issues down the road (ex. including security related ones if the db is stored for some reason in `pb_public`).

**Reply by cavecafe** [reply]
@ganigeorgiev, 
I do not know how the watching the pb_hooks logic is implemented. It is just purely my guess out of thin air.

* If pb_hooks directory is being monitored by 'any file' changes in the folder, it might be something need to be re-considered. Because, I see 'lost+found' at the early state of 'poketbase' runs (play the original GIF image that I posted), which might be generated by something else (not by poketbase, but by other volume mounting methods for PV)

### Good News 
Like I mentioned, simply let 'pocketbase' create these folders on K8s, I see the same result of my local macOS result.

#### For my test: 
* no changes in Dockerfile, exception of removing of the flag (--hooksWatch=false), use as default
```dockerfile
...

WORKDIR /pocketbase

ENTRYPOINT ["/usr/local/bin/pocketbase", "serve", "--http=0.0.0.0:8090", \
    "--dir=/pocketbase/pb_data", \
    "--publicDir=/pocketbase/pb_public", \
    "--hooksDir=/pocketbase/pb_hooks"]
...

```
  
* removed the parts to create sub-directories on my deployment YML file, only prepare base path 
```yaml
          volumeMounts:
            - name: pocketbase
              mountPath: "/pocketbase"
``` 
 No more crashes and restart. 👍 
 Please let me know if you want me to do the further investigation. 
Thanks!

**Reply by thangngoc89** [reply]
@cavecafe `pb_public` and `pb_hooks` should be in the Dockerfile and checked in git if possible. Especially `pb_hooks`. Because pb_hooks is considered as code .
